File: templates/Features/BackgroundTaskVB/BackgroundTasks/BackgroundTask.vb
Imports System.Threading.Tasks

Imports Windows.ApplicationModel.Background

Namespace BackgroundTasks
    Public MustInherit Class BackgroundTask
        Public MustOverride Sub Register()

        Public MustOverride Function RunAsyncInternal(taskInstance As IBackgroundTaskInstance) As Task

        Public MustOverride Sub OnCanceled(sender As IBackgroundTaskInstance, reason As BackgroundTaskCancellationReason)

        Public Function Match(name As String) As Boolean
            Return name = [GetType]().Name
        End Function

        Public Function RunAsync(taskInstance As IBackgroundTaskInstance) As Task
            SubscribeToEvents(taskInstance)

            Return RunAsyncInternal(taskInstance)
        End Function

        Public Sub SubscribeToEvents(taskInstance As IBackgroundTaskInstance)
            AddHandler taskInstance.Canceled, New BackgroundTaskCanceledEventHandler(AddressOf OnCanceled)
        End Sub
    End Class
End Namespace

File: code/TemplateStudioForUWP/Templates/Uwp/Features/BackgroundTask/Param_ProjectName/BackgroundTasks/BackgroundTask.cs
using System;
using System.Threading.Tasks;

using Windows.ApplicationModel.Background;

namespace Param_RootNamespace.BackgroundTasks
{
    public abstract class BackgroundTask
    {
        public abstract void Register();

        public abstract Task RunAsyncInternal(IBackgroundTaskInstance taskInstance);

        public abstract void OnCanceled(IBackgroundTaskInstance sender, BackgroundTaskCancellationReason reason);

        public bool Match(string name)
        {
            return name == GetType().Name;
        }

        public Task RunAsync(IBackgroundTaskInstance taskInstance)
        {
            SubscribeToEvents(taskInstance);

            return RunAsyncInternal(taskInstance);
        }

        public void SubscribeToEvents(IBackgroundTaskInstance taskInstance)
        {
            taskInstance.Canceled += new BackgroundTaskCanceledEventHandler(OnCanceled);
        }
    }
}
